[Screen Time Refactoring] Don't spin up Screen Time in offscreen web views

jesxilin opened the pull request:
#### eb2cf5ec637d91ca194ee664da7ccc1127cc6167
<pre>
[Screen Time Refactoring] Don&apos;t spin up Screen Time in offscreen web views
<a href="https://bugs.webkit.org/show_bug.cgi?id=289784">https://bugs.webkit.org/show_bug.cgi?id=289784</a>
<a href="https://rdar.apple.com/146809373">rdar://146809373</a>

Reviewed by Aditya Keerthi.

Check and do not create the ScreenTimeWebpageController if the web view
is never in window. Also, add an API test for this case.

Change the API test (DoNotDonateURLsInOccludedWebView) which used to
remove the webview from window, then add the webview back to window.
Now, instead of removing the webview from the window, we add a new window
to cover the webview, effectively occluding the webview. Then check for
donations again after making the covering window have zero width and height.

Add an API test (CreateControllerAfterOffscreenWebViewBecomesInWindow) for
the case of an offscreen webview that becomes an onscreen webview.

Add an API test (ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)
to check if we set URLs in the case of an offscreen webview that becomes
an onscreen webview.

Update the API test (ScreenTimeControllerInstalledAfterRestoreFromSessionState)
to not be added to test window upon creation and added after verifying the
controller does not exist yet.

Rename function from _updateScreenTimeShieldVisibilityForWindow
to _updateScreenTimeBasedOnWindowVisibility since we are generally setting
Screen Time information based on the window&apos;s visibility. Additionally,
install the controller if the webview is now in window and we have not
installed the controller previously. Also, set the URL on the controller
if we are installing the controller from _updateScreenTimeBasedOnWindowVisibility.

Rename function from _installScreenTimeWebpageController to
_installScreenTimeWebpageControllerIfNeeded since this is
conditional on whether we need to install the controller or not.

Note: The uninstallation of the controller for
the case of a onscreen webview, then offscreen webview will be
in a separate PR to keep things concise and organized.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _updateScreenTimeBasedOnWindowVisibility]):
(-[WKWebView _installScreenTimeWebpageController]): Deleted.
(-[WKWebView _updateScreenTimeShieldVisibilityForWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::viewIsBecomingVisible):
(WebKit::PageClientImplCocoa::viewIsBecomingInvisible):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, DoNotDonateURLsInOccludedWebView)):
(TEST(ScreenTime, CreateControllerAfterOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)): Deleted.

Canonical link: <a href="https://commits.webkit.org/292400@main">https://commits.webkit.org/292400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78a2e3326b0a742df0c908de4da1038e0f538d70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100980 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97966 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23967 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73135 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30356 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11847 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53459 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4390 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45765 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103009 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22988 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82175 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81538 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26126 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3575 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16325 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15437 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22951 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26090 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -683,6 +683,22 @@ - (void)userContentController:(WKUserContentController *)userContentController d
     EXPECT_FALSE(suppressUsageRecording);
 }
 
+TEST(ScreenTime, DoNotCreateControllerWhenWebViewAlwaysOffscreen)
```
What behavior do you want for web view that becomes off screen at some indeterminate point? You would still have an `STWebPageController` around, is that ok?

pxlcoder replied:
At first, I thought it would be okay to keep it around since donations are suppressed in off screen web views anyway. But on second thought, the XPC connection would be unnecessarily kept alive, which is a kind of memory leak.

Due to that, we should uninstall the controller if the web view goes offscreen.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -335,11 +335,14 @@
         return;
 
-    if (![webView _screenTimeWebpageController])
+    if (![webView _screenTimeWebpageController] && !![webView window])
```
Adding this logic in `updateScreenTimeWebpageControllerURL` doesn't look right.

If a web view is offscreen, URL loaded, and then the web view is added to the window – what installs the controller?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -667,21 +636,51 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+
+    [webView waitForNextPresentationUpdate];
+
+    RetainPtr controller = [webView _screenTimeWebpageController];
```
This variable appears unused / only written to and not read from.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -667,21 +636,51 @@ - (void)userContentController:(WKUserContentController *)userContentController d
-    TestWebKitAPI::Util::run(&done);
+    RetainPtr view = adoptNS([ [NSView alloc] initWithFrame:[webView frame] ]);
+    [view setWantsLayer:YES];
+    [[view layer] setBackgroundColor:NSColor.redColor.CGColor];
```
Don't think this is necessary.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -667,21 +636,51 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    [[view layer] setBackgroundColor:NSColor.redColor.CGColor];
+
+    [[window contentView] addSubview:view.get()];
+    [window contentView].wantsLayer = YES;
```
Probably not necessary either.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -667,21 +636,51 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    RetainPtr window = adoptNS([[NSWindow alloc] initWithContentRect:[webView frame] styleMask:NSWindowStyleMaskBorderless backing:NSBackingStoreBuffered defer:NO]);
 
-    TestWebKitAPI::Util::run(&done);
+    RetainPtr view = adoptNS([ [NSView alloc] initWithFrame:[webView frame] ]);
```
(we probably don't need the view in general, it was more for debugging)

jesxilin replied:
Don't we need the view for `[window makeKeyAndOrderFront:view.get()];`?

pxlcoder replied:
Passing in `nil` should be fine.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,35 +455,20 @@ - (void)_updateScreenTimeViewGeometry
-        }
-    } else {
-        if (_screenTimeBlurredSnapshot) {
-            [_screenTimeBlurredSnapshot setHidden:YES];
```
What updates the blur view now?

pxlcoder replied:
Realized that this was never necessary, since if the view is not in window, it wouldn't be visible anyway and neither would the blur.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,35 +455,20 @@ - (void)_updateScreenTimeViewGeometry
     [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !viewIsVisible)];
+
+    if (viewIsInWindow)
+        [self _installScreenTimeWebpageController];
```
We should probably rename `_installScreenTimeWebpageController` to `_installScreenTimeWebpageControllerIfNeeded` now.

pxlcoder replied:
Also, I think this code should be before the call to `setSuppressUsageRecording`? Otherwise you could have a `nil` `_screenTimeWebpageController`, skipping the suppression, and then it becomes non-nil after this code runs.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -667,21 +636,51 @@ - (void)userContentController:(WKUserContentController *)userContentController d
     done = false;
 
-    [webView addToTestWindow];
+    [window setFrame:CGRectMake(0, 0, 0, 0) display:YES];
```
```suggestion
    [window setFrame:CGRectZero display:YES];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,31 +455,16 @@ - (void)_updateScreenTimeViewGeometry
-    BOOL viewIsVisible = viewIsInWindow;
 #if PLATFORM(MAC)
     viewIsVisible &= ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
 #endif
```
I think this logic should continue to be kept next to `viewIsVisible` (you can move it up).

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,34 +455,20 @@ - (void)_updateScreenTimeViewGeometry
+    else
+        [self _uninstallScreenTimeWebpageController];
+
     [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !viewIsVisible)];
```
```suggestion
    if (viewIsInWindow) {
        [self _installScreenTimeWebpageControllerIfNeeded];
        [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !viewIsVisible)];
    } else
        [self _uninstallScreenTimeWebpageController];
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,34 +455,20 @@ - (void)_updateScreenTimeViewGeometry
     BOOL viewIsVisible = viewIsInWindow;
+
 #if PLATFORM(MAC)
     viewIsVisible &= ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```
FYI this is doing a bitwise AND operation. Better to avoid the ambiguity because I know you want logical AND instead.

```suggestion
    viewIsVisible = viewIsVisible && ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```

aprotyas commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -335,11 +335,14 @@
+
     RefPtr pageProxy = [webView _page].get();
     if (pageProxy && !pageProxy->preferences().screenTimeEnabled()) {
         [webView _uninstallScreenTimeWebpageController];
```
`_installScreenTimeWebpageControllerIfNeeded` already respects the `ScreenTimeEnabled` preference, right? It feels superfluous to then have this flow? This might be easier

```cpp
void PageClientImplCocoa::updateScreenTimeWebpageControllerURL(WKWebView *webView)
{
    if (!PAL::isScreenTimeFrameworkAvailable())
        return;

    if (![webView _screenTimeWebpageController] && !![webView window])
        [webView _installScreenTimeWebpageControllerIfNeeded];

    if (RetainPtr screenTimeWebpageController = [webView _screenTimeWebpageController])
        [screenTimeWebpageController setURL:[webView _mainFrameURL]];
```

pxlcoder replied:
```
if (RetainPtr screenTimeWebpageController = [webView _screenTimeWebpageController])
        [screenTimeWebpageController setURL:[webView _mainFrameURL]];
```

is unnecessary in Obj-C. Calling a method on `nil` does nothing.

aprotyas replied:
oh yes, that too!

aprotyas replied:
```cpp
void PageClientImplCocoa::updateScreenTimeWebpageControllerURL(WKWebView *webView)
{
    if (!PAL::isScreenTimeFrameworkAvailable())
        return;

    if (![webView _screenTimeWebpageController] && !![webView window])
        [webView _installScreenTimeWebpageControllerIfNeeded];

    [[webView  _screenTimeWebpageController] setURL:[webView _mainFrameURL]];
```

pxlcoder replied:
I think the idea here was to uninstall if someone toggles the pref? In which case, maybe we just need the other logic at the top, with an early return.

aprotyas replied:
oh, I see it now.

jesxilin replied:
I'll condense the logic

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,12 +455,16 @@ - (void)_updateScreenTimeViewGeometry
     BOOL showsSystemScreenTimeBlockingView = [_configuration _showsSystemScreenTimeBlockingView];
 
+#if PLATFORM(MAC)
+    viewIsVisible = viewIsVisible && ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```
We should keep `viewIsVisible` logic next to each other? Seems weird to have `BOOL showsSystemScreenTimeBlockingView` in the middle.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -335,17 +335,16 @@
+        [webView _installScreenTimeWebpageControllerIfNeeded];
 
     RefPtr pageProxy = [webView _page].get();
     if (pageProxy && !pageProxy->preferences().screenTimeEnabled()) {
         [webView _uninstallScreenTimeWebpageController];
         return;
     }
```
Per the earlier comment, this should still be moved above right?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -402,7 +402,7 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
-- (void)_installScreenTimeWebpageController
+- (void)_installScreenTimeWebpageControllerIfNeeded
 {
     if (!PAL::isScreenTimeFrameworkAvailable())
```
If you add:

```
if (_screenTimeWebpageController)
    return;
```

here, then you can remove the if on 413, and...

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -335,17 +335,16 @@
         return;
     }
 
+    if (![webView _screenTimeWebpageController] && !![webView window])
```
... simplify this to

```
if ([webView window])
    [webView _installScreenTimeWebpageControllerIfNeeded];
```

jesxilin replied:
Will address both comments!

pxlcoder reviewed (commented):
Does anything set the URL on the `STWebpageController` when going from offscreen to in window?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -402,34 +402,35 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
-        [_screenTimeWebpageController addObserver:self forKeyPath:@"URLIsBlocked" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&screenTimeWebpageControllerBlockedKVOContext];
-        _isBlockedByScreenTime = NO;
+    if (_screenTimeWebpageController)
+        return;
```
This is better at the top, to avoid the marginally more expensive calls.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,13 +456,19 @@ - (void)_updateScreenTimeViewGeometry
 
     if (viewIsInWindow) {
+        [self _installScreenTimeWebpageControllerIfNeeded];
+        [_screenTimeWebpageController setURL:[self _mainFrameURL]];
```
I think we should only do this if a controller was installed.

jesxilin committed:
[Screen Time Refactoring] Don't spin up Screen Time in offscreen web views
https://bugs.webkit.org/show_bug.cgi?id=289784
rdar://146809373

Reviewed by Aditya Keerthi.

Check and do not create the ScreenTimeWebpageController if the web view
is never in window. Also, add an API test for this case.

Change the API test (DoNotDonateURLsInOccludedWebView) which used to
remove the webview from window, then add the webview back to window.
Now, instead of removing the webview from the window, we add a new window
to cover the webview, effectively occluding the webview. Then check for
donations again after making the covering window have zero width and height.

Add an API test (CreateControllerAfterOffscreenWebViewBecomesInWindow) for
the case of an offscreen webview that becomes an onscreen webview.

Add an API test (ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)
to check if we set URLs in the case of an offscreen webview that becomes
an onscreen webview.

Update the API test (ScreenTimeControllerInstalledAfterRestoreFromSessionState)
to not be added to test window upon creation and added after verifying the
controller does not exist yet.

Rename function from _updateScreenTimeShieldVisibilityForWindow
to _updateScreenTimeBasedOnWindowVisibility since we are generally setting
Screen Time information based on the window's visibility. Additionally,
install the controller if the webview is now in window and we have not
installed the controller previously. Also, set the URL on the controller
if we are installing the controller from _updateScreenTimeBasedOnWindowVisibility.

Rename function from _installScreenTimeWebpageController to
_installScreenTimeWebpageControllerIfNeeded since this is
conditional on whether we need to install the controller or not.

Note: The uninstallation of the controller for
the case of a onscreen webview, then offscreen webview will be
in a separate PR to keep things concise and organized.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _updateScreenTimeBasedOnWindowVisibility]):
(-[WKWebView _installScreenTimeWebpageController]): Deleted.
(-[WKWebView _updateScreenTimeShieldVisibilityForWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::viewIsBecomingVisible):
(WebKit::PageClientImplCocoa::viewIsBecomingInvisible):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, DoNotDonateURLsInOccludedWebView)):
(TEST(ScreenTime, CreateControllerAfterOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerSetsURLWhenOffscreenWebViewBecomesInWindow)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)): Deleted.

Canonical link: https://commits.webkit.org/292400@main
